designs/2020-es-module-support/README.md

```diff
@@ -0,0 +1,159 @@
+- Repo: eslint/espree
+- Start Date: 2020-12-13
+- RFC PR: `TODO`
+- Authors: Mike Reinstein (https://github.com/mreinstein)
+
+# Adding es module support
+
+## Summary
+
+Provide an ecmascript module for espree.
+
+
+## Motivation
+
+Historically there have been a number of competing module standards: commonjs, requireJs, iife wrapped script tags, etc.
+
+This has contributed to a very complicated tooling scenario for node/npm/javascript based projects in general:
+* every module has it's own unique build system, with no consistency across modules.
+* projects are packed with complex configurations, much of which is plumbing related to packaging/bundling.
+
+Over time, browsers have adopted the ecma module standard. Today, support is pretty good:
+https://caniuse.com/es6-module
+
+Node also added support in v12.17.1 and up, and has officially marked the API stable:
+https://nodejs.org/dist/latest-v15.x/docs/api/esm.html#esm_modules_ecmascript_modules
+
+
+By adopting this new standard, we can achieve a few things:
+
+* eventual simplification of espree's build process by eliminating the need for multiple module formats
+* make it easier to directly importing espree in browser environments, deno, etc. Skypack is a CDN that already enables this, but they have internal logic that looks at every package in npm and tries to build an es module. https://www.skypack.dev/
+
+As more projects adopt this standardized module format, the hope is a lot of this "connective glue" will fall away, leaving us with leaner modules and less build tooling across our stack.
+
+
+## Detailed Design
+
+### Nomenclature
+
+* `ESM` ecmascript module. The standardized module format for javascript
+* `CJS` commonjs. Node's historical module format (i.e., `module.exports = ...`, `const a = require('foo')`)
+* `UMD` universal module definition. A legacy format for specifying javascript code as a module in a way that tries to guarantee compatibility with many loader formats including requireJS, commonjs, and global script tags.
+* `IIFE` immediately invoked function expressions. A legacy strategy for encapsulating javascript into a module that won't leak global variables into the execution environment. It is still the primary way of packaging global script tags for inclusion in a web page. `UMD` is usually wrapped in an `IIFE`
+* `CDN` content delivery network. can be thought of as a set of servers that host static content all around the globe to deliver assets quickly and at scale.
+
+
+### build process
+
+Today, espree is written in commonjs (CJS) format, and a universal module definition (UMD) build is created too.
```

Comment:
There is a script to create UMD, but are we publishing this bundle or using it to run any tests? If neither of those, maybe we should just drop it.

Reply:
I can’t think of a good reason to generate UMD. I’m in favor of dropping it.

Reply:
I *think* what you're referring to here is the browserify step we have in the eslint repo? We use that just to create a browser-safe version of eslint for our demo and isn't an officially supported version of Espree. (You can't reference a UMD version from the Espree package.)